index.js

```diff
@@ -1,284 +1,161 @@
-    else if (tf.equals(BOX2_TF)) return d.toString('base64') + '.box2'
-    else throw new Error('Unknown box: ' + buf)
-  },
-
-  string(buf) {
-    const d = buf.slice(2)
-    return d.toString('utf8')
-  },
-
-  boolean(buf) {
-    const d = buf.slice(2)
-    return d.equals(BOOL_TRUE)
+  string(input) {
+    return input.slice(2).toString('utf8')
   },
 }
 
 function decode(input) {
-  if (Array.isArray(input)) {
-    return input.map(decode)
-  } else if (Buffer.isBuffer(input)) {
+  /* cases we don't decode */
+  if (input === null) return null
+  if (Number.isInteger(input)) return input
+
+  if (Buffer.isBuffer(input)) {
     if (input.length < 2)
       throw new Error(
         'Buffer is missing first two type&format fields: ' + input
       )
-    const t = input.slice(0, 1)
-    const tf = input.slice(0, 2)
-    if (tf.equals(STRING_TF)) return decoder.string(input)
-    else if (tf.equals(BOOL_TF)) return decoder.boolean(input)
-    else if (tf.equals(NIL_TF)) return null
-    else if (t.equals(FEED_T)) return decoder.feed(input)
-    else if (t.equals(MSG_T)) return decoder.message(input)
-    else if (t.equals(BLOB_T)) return decoder.blob(input)
-    else if (t.equals(BOX_T)) return decoder.box(input)
-    else if (t.equals(SIGNATURE_T)) {
-      if (tf.equals(SIGNATURE_TF)) return decoder.signature(input)
-      else throw new Error('Unknown signature type')
-    } else if (tf.equals(SIGNATURE_TF)) return decoder.signature(input)
-    else return input
-  } else if (typeof input === 'object' && input !== null) {
+
+    if (input.equals(NIL_TF)) return null
+    if (input.slice(0, 2).equals(BOOL.TF)) return decoder.bool(input)
+    if (input.slice(0, 2).equals(STRING_TF)) return decoder.string(input)
+
+    const type = TYPES.find((type) => type.code.equals(input.slice(0, 1)))
+    if (type) {
+      const format = type.formats.find((format) =>
+        format.code.equals(input.slice(1, 2))
+      )
+      if (format) {
+        if (type.sigil || format.suffix)
+          return decoder.sigilSuffix(input, type, format)
+        else
+          throw new Error(
+            `no decoder defined yet for type: ${type.type}, format: ${format.format}`
+          )
+      } else throw new Error(`Unknown ${type.type} format`)
+    }
+
+    throw new Error('Unknown type/ format')
+  }
+
+  /* recurse */
+  if (Array.isArray(input)) return input.map(decode)
+  if (typeof input === 'object') {
+    // know it's not null, Array
     const output = {}
-    for (let key in input) {
-      const y = input[key]
-      const x = decode(y)
-      output[key] = x
+    for (const key in input) {
+      output[key] = decode(input[key])
     }
     return output
-  } // FIXME: more checks, including floats!
-  else return input
+  }
+
+  // FIXME: more checks, including floats!
+  throw new Error("don't know how to decode: " + input)
 }
 
 module.exports = {
   encode,
   decode,
-  bfeTypes: TYPES,
+  toString: decode, // alias
```

Comment:
`toString` is highly misleading because the decoder can return booleans, null, and buffers. I suggest removing this.

Reply:
quite right.
removed